File: backend/app/schemas/mission.py
from datetime import datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, Field, field_validator

from app.schemas.common import ListMeta
from app.schemas.geometry import PointZ

# computation status values - mirrors ComputationStatus enum
ComputationStatusStr = Literal["IDLE", "COMPUTING", "COMPLETED", "FAILED"]

# flight plan scope values - mirrors FlightPlanScope enum
FlightPlanScopeStr = Literal["FULL", "NO_TAKEOFF_LANDING", "MEASUREMENTS_ONLY"]

# inspection method values - mirrors InspectionMethod enum
InspectionMethodStr = Literal[
    "VERTICAL_PROFILE",
    "PAPI_HORIZONTAL_RANGE",
    "FLY_OVER",
    "PARALLEL_SIDE_SWEEP",
    "HOVER_POINT_LOCK",
]
# capture mode values - used by trajectory_computation to choose camera_action
CaptureModeStr = Literal["VIDEO_CAPTURE", "PHOTO_CAPTURE"]
# hover bearing reference frames - RUNWAY = 0 is approach side, COMPASS = absolute
HoverBearingRefStr = Literal["RUNWAY", "COMPASS"]

# minimum transit altitude (m AGL) - mirrors trajectory_types.MINIMUM_AGL_ALTITUDE.
# duplicated here so schemas do not import from services (architectural boundary).
_MIN_TRANSIT_ALTITUDE_AGL = 5.0


def _validate_transit_altitude(value: float | None) -> float | None:
    """enforce transit altitude minimum without importing from services."""
    if value is None:
        return None
    if value < _MIN_TRANSIT_ALTITUDE_AGL:
        raise ValueError(f"transit_agl must be at least {_MIN_TRANSIT_ALTITUDE_AGL:.0f}m AGL")
    return value


class InspectionConfigOverride(BaseModel):
    """config overrides for an inspection within a mission"""

    altitude_offset: float | None = None
    measurement_speed_override: float | None = Field(default=None, gt=0)
    measurement_density: int | None = Field(default=None, ge=1)
    custom_tolerances: dict[str, float] | None = None
    hover_duration: float | None = None
    horizontal_distance: float | None = Field(default=None, gt=0)
    sweep_angle: float | None = None
    vertical_profile_height: float | None = Field(default=None, gt=0)
    lha_ids: list[UUID] | None = None
    capture_mode: CaptureModeStr | None = None
    recording_setup_duration: float | None = None
    buffer_distance: float | None = Field(default=None, ge=0)
    # new method-specific config fields
    height_above_lights: float | None = Field(default=None, gt=0)
    lateral_offset: float | None = Field(default=None, gt=0)
    distance_from_lha: float | None = Field(default=None, gt=0)
    height_above_lha: float | None = Field(default=None, gt=0)
    camera_gimbal_angle: float | None = None
    selected_lha_id: UUID | None = None
    hover_bearing: float | None = None
    hover_bearing_reference: HoverBearingRefStr | None = None

    @field_validator("lha_ids", mode="before")
    @classmethod
    def validate_lha_ids(cls, v: list | None) -> list[UUID] | None:
        """coerce mixed uuid/string lists so downstream jsonb storage is consistent."""
        if v is None:
            return None
        return [UUID(str(i)) if not isinstance(i, UUID) else i for i in v]


class InspectionCreate(BaseModel):
    """add inspection to mission"""

    template_id: UUID
    method: InspectionMethodStr
    config: InspectionConfigOverride | None = None


class InspectionUpdate(BaseModel):
    """update inspection within mission"""

    method: InspectionMethodStr | None = None
    config: InspectionConfigOverride | None = None
    sequence_order: int | None = None


class InspectionConfigResponse(BaseModel):
    """inspection configuration values"""

    altitude_offset: float | None = None
    measurement_speed_override: float | None = None
    measurement_density: int | None = None
    custom_tolerances: dict[str, float] | None = None
    hover_duration: float | None = None
    horizontal_distance: float | None = None
    sweep_angle: float | None = None
    vertical_profile_height: float | None = None
    lha_ids: list[UUID] | None = None
    capture_mode: CaptureModeStr | None = None
    recording_setup_duration: float | None = None
    buffer_distance: float | None = None
    height_above_lights: float | None = None
    lateral_offset: float | None = None
    distance_from_lha: float | None = None
    height_above_lha: float | None = None
    camera_gimbal_angle: float | None = None
    selected_lha_id: UUID | None = None
    hover_bearing: float | None = None
    hover_bearing_reference: HoverBearingRefStr | None = None

    model_config = {"from_attributes": True}


class InspectionResponse(BaseModel):
    """inspection response"""

    id: UUID
    mission_id: UUID
    template_id: UUID
    config_id: UUID | None = None
    method: InspectionMethodStr
    sequence_order: int
    lha_ids: list[UUID] | None = None
    config: InspectionConfigResponse | None = None

    model_config = {"from_attributes": True}


class ReorderRequest(BaseModel):
    """reorder inspections by sequence"""

    inspection_ids: list[UUID]


class ReorderResponse(BaseModel):
    """reorder response"""

    reordered: bool


class MissionCreate(BaseModel):
    """create mission"""

    name: str
    airport_id: UUID
    drone_profile_id: UUID | None = None
    operator_notes: str | None = None
    default_speed: float | None = None
    measurement_speed_override: float | None = Field(default=None, gt=0)
    default_altitude_offset: float | None = None
    takeoff_coordinate: PointZ | None = None
    landing_coordinate: PointZ | None = None
    default_capture_mode: CaptureModeStr | None = None
    default_buffer_distance: float | None = Field(default=None, ge=0)
    transit_agl: float | None = None
    require_perpendicular_runway_crossing: bool = True
    flight_plan_scope: FlightPlanScopeStr = "FULL"

    @field_validator("transit_agl")
    @classmethod
    def _check_transit_altitude(cls, v: float | None) -> float | None:
        """enforce minimum AGL floor on mission-level cruise altitude."""
        return _validate_transit_altitude(v)


class MissionUpdate(BaseModel):
    """update mission"""

    name: str | None = None
    drone_profile_id: UUID | None = None
    operator_notes: str | None = None
    default_speed: float | None = None
    measurement_speed_override: float | None = Field(default=None, gt=0)
    default_altitude_offset: float | None = None
    takeoff_coordinate: PointZ | None = None
    landing_coordinate: PointZ | None = None
    date_time: datetime | None = None
    default_capture_mode: CaptureModeStr | None = None
    default_buffer_distance: float | None = Field(default=None, ge=0)
    transit_agl: float | None = None
    require_perpendicular_runway_crossing: bool | None = None
    flight_plan_scope: FlightPlanScopeStr | None = None

    @field_validator("transit_agl")
    @classmethod
    def _check_transit_altitude(cls, v: float | None) -> float | None:
        """enforce minimum AGL floor on mission-level cruise altitude."""
        return _validate_transit_altitude(v)


class MissionResponse(BaseModel):
    """mission response"""

    id: UUID
    name: str
    status: str
    airport_id: UUID
    created_at: datetime
    updated_at: datetime
    operator_notes: str | None = None
    drone_profile_id: UUID | None = None
    date_time: datetime | None = None
    default_speed: float | None = None
    measurement_speed_override: float | None = None
    default_altitude_offset: float | None = None
    takeoff_coordinate: PointZ | None = None
    landing_coordinate: PointZ | None = None
    default_capture_mode: CaptureModeStr | None = None
    default_buffer_distance: float | None = None
    transit_agl: float | None = None
    require_perpendicular_runway_crossing: bool = True
    flight_plan_scope: FlightPlanScopeStr = "FULL"
    has_unsaved_map_changes: bool = False
    computation_status: ComputationStatusStr = "IDLE"
    computation_error: str | None = None
    computation_started_at: datetime | None = None
    inspection_count: int = 0
    estimated_duration: float | None = None

    model_config = {"from_attributes": True}


class MissionDetailResponse(MissionResponse):
    """mission with inspections"""

    inspections: list[InspectionResponse] = []


class ComputationStatusResponse(BaseModel):
    """lightweight computation status for polling."""

    computation_status: ComputationStatusStr
    computation_error: str | None = None
    computation_started_at: datetime | None = None

    model_config = {"from_attributes": True}


class MissionListResponse(BaseModel):
    """mission list response"""

    data: list[MissionResponse]
    meta: ListMeta

File: frontend/src/types/mission.ts
import type { PointZ } from "./common";
import type { CaptureMode, ComputationStatus, FlightPlanScope, InspectionMethod, MissionStatus } from "./enums";

export type CameraMode = "AUTO" | "MANUAL";

export interface MissionResponse {
  id: string;
  name: string;
  status: MissionStatus;
  airport_id: string;
  created_at: string;
  updated_at: string;
  operator_notes: string | null;
  drone_profile_id: string | null;
  date_time: string | null;
  default_speed: number | null;
  measurement_speed_override: number | null;
  default_altitude_offset: number | null;
  takeoff_coordinate: PointZ | null;
  landing_coordinate: PointZ | null;
  default_capture_mode: CaptureMode | null;
  default_buffer_distance: number | null;
  camera_mode: CameraMode;
  default_white_balance: string | null;
  default_iso: number | null;
  default_shutter_speed: string | null;
  default_focus_mode: "AUTO" | "INFINITY" | null;
  transit_agl: number | null;
  require_perpendicular_runway_crossing: boolean;
  flight_plan_scope: FlightPlanScope;
  has_unsaved_map_changes: boolean;
  computation_status: ComputationStatus;
  computation_error: string | null;
  computation_started_at: string | null;
  inspection_count: number;
  estimated_duration: number | null;
}

export interface MissionDetailResponse extends MissionResponse {
  inspections: InspectionResponse[];
}

export interface InspectionConfigResponse {
  altitude_offset: number | null;
  angle_offset: number | null;
  measurement_speed_override: number | null;
  measurement_density: number | null;
  custom_tolerances: Record<string, number> | null;
  hover_duration: number | null;
  horizontal_distance: number | null;
  sweep_angle: number | null;
  vertical_profile_height: number | null;
  lha_ids: string[] | null;
  capture_mode: CaptureMode | null;
  recording_setup_duration: number | null;
  buffer_distance: number | null;
  height_above_lights: number | null;
  lateral_offset: number | null;
  distance_from_lha: number | null;
  height_above_lha: number | null;
  camera_gimbal_angle: number | null;
  selected_lha_id: string | null;
  lha_setting_angle_override_id: string | null;
  hover_bearing: number | null;
  hover_bearing_reference: "RUNWAY" | "COMPASS" | null;
  camera_mode: CameraMode | null;
  direction_reversed: boolean;
  direction_is_auto: boolean;
  white_balance: string | null;
  iso: number | null;
  shutter_speed: string | null;
  focus_mode: "AUTO" | "INFINITY" | null;
  optical_zoom: number | null;
  camera_preset_id: string | null;
}

export interface InspectionResponse {
  id: string;
  mission_id: string;
  template_id: string;
  config_id: string | null;
  method: InspectionMethod;
  sequence_order: number;
  lha_ids: string[] | null;
  config: InspectionConfigResponse | null;
}

export interface InspectionConfigOverride {
  altitude_offset?: number | null;
  angle_offset?: number | null;
  measurement_speed_override?: number | null;
  measurement_density?: number | null;
  custom_tolerances?: Record<string, number> | null;
  hover_duration?: number | null;
  horizontal_distance?: number | null;
  sweep_angle?: number | null;
  vertical_profile_height?: number | null;
  lha_ids?: string[] | null;
  capture_mode?: CaptureMode | null;
  recording_setup_duration?: number | null;
  buffer_distance?: number | null;
  height_above_lights?: number | null;
  lateral_offset?: number | null;
  distance_from_lha?: number | null;
  height_above_lha?: number | null;
  camera_gimbal_angle?: number | null;
  selected_lha_id?: string | null;
  lha_setting_angle_override_id?: string | null;
  hover_bearing?: number | null;
  hover_bearing_reference?: "RUNWAY" | "COMPASS" | null;
  camera_mode?: CameraMode | null;
  direction_reversed?: boolean;
  direction_is_auto?: boolean;
  white_balance?: string | null;
  iso?: number | null;
  shutter_speed?: string | null;
  focus_mode?: "AUTO" | "INFINITY" | null;
  optical_zoom?: number | null;
  camera_preset_id?: string | null;
}

export interface MissionCreate {
  name: string;
  airport_id: string;
  operator_notes?: string | null;
  drone_profile_id?: string | null;
  date_time?: string | null;
  default_speed?: number | null;
  measurement_speed_override?: number | null;
  default_altitude_offset?: number | null;
  takeoff_coordinate?: PointZ | null;
  landing_coordinate?: PointZ | null;
  default_capture_mode?: CaptureMode | null;
  default_buffer_distance?: number | null;
  camera_mode?: CameraMode;
  default_white_balance?: string | null;
  default_iso?: number | null;
  default_shutter_speed?: string | null;
  default_focus_mode?: "AUTO" | "INFINITY" | null;
  transit_agl?: number | null;
  require_perpendicular_runway_crossing?: boolean;
  flight_plan_scope?: FlightPlanScope;
}

export interface MissionUpdate {
  name?: string;
  operator_notes?: string | null;
  drone_profile_id?: string | null;
  date_time?: string | null;
  default_speed?: number | null;
  measurement_speed_override?: number | null;
  default_altitude_offset?: number | null;
  takeoff_coordinate?: PointZ | null;
  landing_coordinate?: PointZ | null;
  default_capture_mode?: CaptureMode | null;
  default_buffer_distance?: number | null;
  camera_mode?: CameraMode;
  default_white_balance?: string | null;
  default_iso?: number | null;
  default_shutter_speed?: string | null;
  default_focus_mode?: "AUTO" | "INFINITY" | null;
  transit_agl?: number | null;
  require_perpendicular_runway_crossing?: boolean;
  flight_plan_scope?: FlightPlanScope;
}

export interface InspectionCreate {
  template_id: string;
  method: InspectionMethod;
  config?: InspectionConfigOverride | null;
}

export interface InspectionUpdate {
  method?: InspectionMethod;
  sequence_order?: number;
  config?: InspectionConfigOverride | null;
}

export interface ReorderRequest {
  inspection_ids: string[];
}

export interface ComputationStatusResponse {
  computation_status: ComputationStatus;
  computation_error: string | null;
  computation_started_at: string | null;
}

export interface HeadingAssignment {
  inspection_id: string;
  sequence_order: number;
  direction_reversed: boolean;
  is_auto: boolean;
}

export interface HeadingAutoResponse {
  mission_id: string;
  assignments: HeadingAssignment[];
  total_distance_m: number;
  total_turn_deg: number;
  baseline_distance_m: number;
  baseline_turn_deg: number;
  improvement_pct: number;
  auto_inspection_count: number;
  pinned_inspection_count: number;
}
